src/arizona_parse_transform.erl

```diff
@@ -186,42 +385,674 @@ transform_stateless_to_ast(StatelessList) when is_list(StatelessList) ->
+    FunctionBindings,
+    CurrentFunctionBindings
+) ->
+    transform_template_calls_with_context(
+        AstNode,
+        ModuleName,
+        CompilerOptions,
+        Depth,
+        ArizonaFunctions,
+        FunctionBindings,
+        CurrentFunctionBindings
+    ).
+
+%% Transform arizona_html function calls with enhanced context
+-spec transform_template_calls_with_context(
+    erl_parse:abstract_expr(),
+    atom(),
+    compile_options(),
+    non_neg_integer(),
+    [function_spec()],
+    #{function_spec() => #{binary() => binary()}},
+    #{binary() => binary()}
+) -> erl_parse:abstract_expr().
+transform_template_calls_with_context(
+    {function, _Line, FuncName, Arity, _Clauses} = Function,
+    _ModuleName,
+    _CompilerOptions,
+    _Depth,
+    ArizonaFunctions,
+    FunctionBindings,
+    _CurrentFunctionBindings
+) ->
+    %% Transform function if it's an Arizona function
+    FuncSpec = {FuncName, Arity},
+    case lists:member(FuncSpec, ArizonaFunctions) of
+        true ->
+            _VarBindings = maps:get(FuncSpec, FunctionBindings, #{}),
+            transform_function_with_bindings(Function);
+        false ->
+            %% Regular transformation for non-Arizona functions
+            Function
+    end;
+transform_template_calls_with_context(
     {call, CallAnnotations,
         {remote, _RemoteAnnotations, {atom, _ModuleAnnotations, arizona_html},
             {atom, _FunctionAnnotations, FunctionName}} = RemoteCall,
         Args},
     ModuleName,
     CompilerOptions,
-    Depth
+    Depth,
+    _ArizonaFunctions,
+    _FunctionBindings,
+    CurrentFunctionBindings
 ) ->
-    transform_arizona_html_call(
-        FunctionName, CallAnnotations, RemoteCall, Args, ModuleName, CompilerOptions, Depth
+    transform_arizona_html_call_with_context(
+        FunctionName,
+        CallAnnotations,
+        RemoteCall,
+        Args,
+        ModuleName,
+        CompilerOptions,
+        Depth,
+        CurrentFunctionBindings
     );
-transform_template_calls(AstNode, _ModuleName, _CompilerOptions, _Depth) ->
+transform_template_calls_with_context(
+    AstNode,
+    _ModuleName,
+    _CompilerOptions,
+    _Depth,
+    _ArizonaFunctions,
+    _FunctionBindings,
+    _CurrentFunctionBindings
+) ->
     AstNode.
 
+%% Transform specific arizona_html function calls with context
+-spec transform_arizona_html_call_with_context(
+    atom(),
+    erl_anno:anno(),
+    erl_parse:abstract_expr(),
+    [erl_parse:abstract_expr()],
+    atom(),
+    compile_options(),
+    non_neg_integer(),
+    #{binary() => binary()}
+) -> erl_parse:abstract_expr().
+transform_arizona_html_call_with_context(
+    render_stateful,
+    CallAnnotations,
+    RemoteCall,
+    Args,
+    ModuleName,
+    _CompilerOptions,
+    _Depth,
+    CurrentFunctionBindings
+) ->
+    %% Use current function bindings for enhanced template processing
+    case CurrentFunctionBindings of
+        EmptyMap when map_size(EmptyMap) =:= 0 ->
```

Comment:
I'm confused by this.

Do you mean simply

```erlang
case CurrentFunctionBindings of
  #{} ->
    ...
```

?

Reply:
That is correct, and again this has bit me today. `#{}` is "map", not "empty map", so my initial comment is irrelevant.